libs/mock-block-dock/src/datastore/traverse.ts

```diff
@@ -1,284 +1,543 @@
+        typeof TIMESTAMP_PLACEHOLDER,
+        KnowledgeGraphOutwardEdge[]
+      >;
+    }
+  >;
+};
+
+export const isTemporalTraversalSubgraph = <
+  RootType extends SubgraphRootType<boolean> = SubgraphRootType<boolean>,
+>(
+  traversalSubgraph: TraversalSubgraph<boolean, RootType>,
+): traversalSubgraph is TraversalSubgraph<true, RootType> => {
+  // this cast is safe as we're only checking against undefined
+  return (
+    (traversalSubgraph as TraversalSubgraph<true>).temporalAxes !== undefined
   );
+};
 
-  if (Object.keys(unresolvedDepths).length === 0) {
-    return;
-  }
+type PatchedOutgoingLinkEdge = PatchEntityValidInterval<OutgoingLinkEdge>;
+type PatchedIncomingLinkEdge = PatchEntityValidInterval<IncomingLinkEdge>;
+type PatchedHasLeftEntityEdge = PatchEntityValidInterval<HasLeftEntityEdge>;
+type PatchedHasRightEntityEdge = PatchEntityValidInterval<HasRightEntityEdge>;
 
-  const element =
-    datastore.vertices[elementIdentifier.baseId]?.[
-      elementIdentifier.revisionId
-    ];
+/**
+ * Parallel of {@link addKnowledgeGraphEdgeToSubgraphByMutation} except that it operates on the patched
+ * {@link TraversalSubgraph} and patched {@link KnowledgeGraphOutwardEdge} types.
+ *
+ * As such, this does not fully qualify the edge at point of insertion; later in the process there is a step which
+ * updates the intervals for entity-related endpoints in the edges to be correct for the {@link Subgraph} as a whole.
+ *
+ * @param {TraversalSubgraph} traversalSubgraph
+ * @param {EntityId} sourceEntityId
+ * @param outwardEdge
+ */
+const patchedAddKnowledgeGraphEdge = <Temporal extends boolean>(
+  traversalSubgraph: TraversalSubgraph<Temporal>,
+  sourceEntityId: EntityId,
+  outwardEdge: PatchEntityValidInterval<KnowledgeGraphOutwardEdge>,
+) =>
+  addKnowledgeGraphEdgeToSubgraphByMutation(
+    // intermediary `as unknown` cast is needed because otherwise tsc gets confused and complains about type
+    // instantiation being excessively deep and possibly infinite
+    traversalSubgraph as unknown as Subgraph<Temporal>,
+    sourceEntityId,
+    TIMESTAMP_PLACEHOLDER,
+    outwardEdge as KnowledgeGraphOutwardEdge,
+  );
 
-  if (!element) {
-    throw new Error(
-      `Couldn't find element in graph associated with identifier: ${JSON.stringify(
-        elementIdentifier,
-      )}`,
-    );
+export const getNeighbors = <
+  EdgeKind extends OutwardEdge["kind"],
+  Reversed extends boolean,
+>(
+  traversalSubgraph: TraversalSubgraph<true>,
+  datastore: Subgraph<true>,
+  source: Vertex<true>,
+  edgeKind: EdgeKind,
+  reversed: Reversed,
+  interval: NonNullTimeInterval,
+): Vertex<true>[] => {
+  switch (edgeKind) {
+    case "HAS_LEFT_ENTITY": {
+      if (!isEntityVertex(source)) {
+        return [];
+      }
+
+      const { inner: sourceEntity } = source;
+
+      const sourceEntityId = sourceEntity.metadata.recordId.entityId;
+
+      if (reversed) {
+        // get outgoing links for entity
```

Comment:
The function name is pretty descriptive, I don't think this comment is needed here as it is. I'd say we either should remove it or give a reason, why we're getting outgoing links, if `"HAS_LEFT_ENTITY"` is reversed.

Reply:
I agree with this, this is a left-over from a previous implementation which didn't have as descriptive function names directly below, have removed